Smt/Solver.lean

```diff
@@ -40,84 +42,197 @@ register_option smt.solver.path : String := {
-def defineConst (id : String) (s : Term) (v : Term) : Solver :=
-  ⟨s!"(define-const {quoteSymbol id} {s} {v})" :: solver.commands⟩
+def defineConst (id : String) (s : Term) (v : Term) : SolverT m Unit :=
+  addCommand (.defineFun id [] s v false)
 
 /-- Declare an uninterpreted function with name `id` and sort `s`. -/
-def declareFun (id : String) (s : Term) : Solver :=
-  ⟨s!"(declare-fun {quoteSymbol id} {s})" :: solver.commands⟩
+def declareFun (id : String) (s : Term) : SolverT m Unit := addCommand (.declare id s)
 
 /-- Define a function with name `id`, parameters `ps`, co-domain `s`,
     and body `t`. -/
 def defineFun (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
-  Solver :=
-  ⟨s!"(define-fun {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
-  where
-    paramsToString (ps : List (String × Term)) : String :=
-      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))
+  SolverT m Unit := addCommand (.defineFun id ps s t false)
 
 /-- Define a recursive function with name `id`, parameters `ps`, co-domain `s`,
     and body `t`. -/
-def defineFunRec (id : String) (ps : List (String × Term)) (s : Term)
-  (t : Term) :
-  Solver :=
-  ⟨s!"(define-fun-rec {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
-  where
-    paramsToString (ps : List (String × Term)) : String :=
-      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))
+def defineFunRec (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
+  SolverT m Unit := addCommand (.defineFun id ps s t true)
 
 /-- Assert that proposition `t` is true. -/
-def assert (t : Term) : Solver :=
-  ⟨s!"(assert {t})" :: solver.commands⟩
+def assert (t : Term) : SolverT m Unit := addCommand (.assert t)
 
 /-- Check if the query given so far is satisfiable and return the result. -/
-def checkSat (kind : Kind) (path : String) : IO String := do
-  let proc ← IO.Process.spawn {
-    stdin := IO.Process.Stdio.piped
-    stdout := IO.Process.Stdio.piped
-    stderr := IO.Process.Stdio.piped
-    cmd := path
-    args := match kind with
-      | Kind.cvc5    => #["-q"]
-      | Kind.z3      => #["-in"]
-      | Kind.vampire => #["--input_syntax", "smtlib2", "--output_mode", "smtcomp"]
-  }
-  let query := String.intercalate
-                 "\n" ("(exit)\n" :: "(check-sat)" :: solver.commands).reverse
-  proc.stdin.putStr query
-  -- Close stdin to signal EOF to the solver.
-  let (_, proc) ← proc.takeStdin
-  _ ← proc.wait
-  proc.stdout.readToEnd
+def checkSat : SolverT m Result := do
+  addCommand .checkSat
+  let out ← getSexp
+  let res ← match out with
+    | "sat"     => return .sat
+    | "unsat"   => return .unsat
+    | "unknown" => return .unknown
+    | _ => (throw (IO.userError s!"unexpected solver output: {repr out}") : IO Result)
+  return res
+
+/- TODO: We should probably parse the returned string into `Command`s or `String × Term`s. This is
```

Comment:
Is `Command` part of the public API to begin with?

Reply:
I would suggest keeping exactly one of (a) the imperative interface with `declareConst` etc as functions in `SolverT` or (b) the data-based `emitCommand` and `Command` as public. I think the latter is nice since it is easier to store data than addresses of methods, but maybe there are other considerations.

Reply:
I think both approaches work and require similar amount of effort, overall. My main reason for switching to a monad is to improve interactivity. Many procedures branch on results of commands like `check-sat` and you probably used `check-sat` + `get-model` loop before to find all solutions to a problem. Doing that is, of course, possible with the data-based approach. But you will probably end up rewriting `SolverT` in the process. My other concern is how easy it would be to extend this API with `cvc5` specific commands. It's not clear to me which approach works better in this case, but I will try my luck with monads first.

Reply:
I wasn't very clear, sorry. I entirely agree that the `SolverT` approach is the way to go. What I mean is just that for the commands that don't return anything other than a possible failure message (you're just sending an assertion/declaration/definition, we could keep `emitCommand (cmd : Command) : SolverT Unit` (or that plus some exception monad) and remove the `declareConst/declareFun/defineFun : SolverT` to avoid duplication. Commands that do return something like `getModel : SolverT Model` must be kept.

Reply:
hmm... I see. Let's keep both approaches for now.